File: 154_find_minimum_in_rotated_sorted_array_ii/Solution.java
class Solution {
    public int minArray(int[] nums) {
        int l = 0, r = nums.length - 1;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (nums[mid] < nums[r]) r = mid;
            else if (nums[mid] > nums[r]) l = mid + 1;
            else r--;
        }

        return nums[l];
    }

    public static void main(String[] args) {
        int[] nums = {3, 4, 5, 1, 2};
        System.out.println(new Solution().minArray(nums));
    }
}

File: 154_find_minimum_in_rotated_sorted_array_ii/Solution.py
class Solution:
    def minArray(self, nums: list) -> int:
        l, r = 0, len(nums) - 1

        while l < r:
            mid = (l + r) // 2
            if nums[mid] < nums[r]: r = mid
            elif nums[mid] > nums[r]: l = mid + 1
            else: r -= 1

        return nums[l]


if __name__ == '__main__':
    nums = [3, 4, 5, 1, 2]
    print(Solution().minArray(nums))
